### ISSUE TITLE
Add async versions of resource loading routines

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#289]
At present, the primary way to load assets in Sphere v2 is using specific overloads of the relevant constructor which accept a filename as input, e.g.

```js
let tex = new Texture('path/to/textureFile.png');
// do stuff with tex
```

This means all assets are loaded synchronously—the constructor doesn’t return until the file is loaded and a valid object has been created from it.  Unfortunately, while this works fine in miniSphere where assets are always available locally and the only event loop the engine needs to answer to is its own, it’s not Web-friendly; in that environment, resources may need to be downloaded on-the-fly and JavaScript code can’t block the browser’s event loop waiting for that to finish.

To solve this problem, any constructor overloads accepting a filename should be deprecated and replaced with Promise-based (`async`) equivalents.  For example, `Texture.fromFile()` would return a promise that resolves to a Texture object.

**Comment by rhuanjl** [comment]
What about instead having the primary means of resource loading being:
1. load the file as an array buffer data THEN
2. use the constructor with the array buffer as the parameter to create the asset

(Reason for this pattern is that it makes the resource objects subclass able whereas a .fromFile method won't work with sub classing)

**Comment by fatcerberus** [comment]
I was thinking about how subclassing could work with this system; it's completely possible for a function like `Texture.fromFile()` to do (the native equivalent of):
```js
let o = Object.create(this.prototype);
await loadTextureFileInto(o, fileName);
return o;
```

But doing so wouldn't call the subclass constructor(s), which might be surprising and/or undesirable.

As for ArrayBuffers, there is still #230 we can implement, but we do still need an async reading function; `FS.readFile()` was unfortunately implemented as sync.  Something like the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch), but without all the bloat, might be good to add.

**Comment by fatcerberus** [comment]
There are now experimental functions such as `Texture.fromFile` in the API which are available for use in miniSphere 5.3b2.  These don't work with subclassing out of the box; they must be extended manually if you need a subclassed version to return the correct object type.

It may still be possible to support the constructors in Oozaru in some capacity, so I won't consider them
deprecated just yet.  But if for whatever reason, you want to ensure that an asset is fully loaded before using it in all implementations, you should now use the `.fromFile()` version and await it rather than calling the constructor.
